### PR TITLE
feat(python): Easier subclassing

### PR DESCRIPTION
**Pull request by ml31415**
Referring to https://github.com/pola-rs/polars/issues/7200 that's a possibility how `DataFrame` and `Series` would be easier to subclass. Previous test cases pass untouched, test cases for subclassing got added, but could still need some extension.

### CODE REVIEW

**Review by stinodego — CHANGES_REQUESTED** [review]
At first glance, I have a bunch of remarks on the implementation side of this PR. But I'd like to see an agreement in the related issue before I do a full review, as it might turn out we do not want this functionality at all.

**Review by stinodego — CHANGES_REQUESTED** [review]
There are a lot of implementation issues with this PR. I don't really know where to start. I left a few comments, but it is not a comprehensive review.

I think the main issue is that this PR promises easier subclassing, but doesn't really deliver on that promise, while adding a lot of clutter and making the code less readable.

The goal of this PR is a noble one. In the existing code base, subclassing sort-of works, but is very limited. Some operations preserve the subclass, while others do not. I recently made some improvements in this regard (see https://github.com/pola-rs/polars/pull/6882), but to be fair, we're not at a point yet where we can say that we offer reliable subclassing support.

If we merge this PR, we would be making some concessions in code readability, but still cannot claim to have a working subclassing implementation. Most importantly, using namespace methods still won't preserve subclasses.

To improve subclassing, the following would be top-of-mind for me:

Make namespaces generic over the Series/Expr they originate from (similar to GroupBy). This allows preserving subclasses in namespace methods.
Update the Series dispatch decorator to preserve subclass.
Address some stray methods that could preserve subclass but currently do not for whatever reason.
Make sure every relevant method returns Self and passes type checking.
Make sure this feature gets documented somewhere.
These are steps we can take without making any concessions in code readability/quality. As it stands, I am not eager to make any concessions since none of us seem too enthousiastic about the pattern described in the linked issue.

File: py-polars/polars/internals/dataframe/frame.py [hunk 1]
```diff
@@ -405,6 +413,10 @@ def __init__(
                 f"DataFrame constructor called with unsupported type; got {type(data)}"
             )
 
+    @classmethod
+    def wrap_df(cls, df: PyDataFrame) -> DataFrame:
+        return cls._from_pydf(df)
+
```
[the remark this answers is not in the document]
Comment (ml31415) [reply py-polars/polars/internals/dataframe/frame.py:419]: It was supposed to serve as a replacement for `wrap_df` on class level, equivalent to the classmethod `wrap_s` for the series.

File: py-polars/polars/internals/series/series.py [hunk 2]
```diff
@@ -422,23 +426,23 @@ def __len__(self) -> int:
...
+        return self.wrap_s(self._s.bitor(other._s))
 
     def __ror__(self, other: Series) -> Series:
         return self.__or__(other)
 
     def __xor__(self, other: Series) -> Series:
         if not isinstance(other, Series):
             other = Series([other])
-        return wrap_s(self._s.bitxor(other._s))
+        return self.wrap_s(self._s.bitxor(other._s))
```
[the remark this answers is not in the document]
Comment (ml31415) [reply py-polars/polars/internals/series/series.py:445]: As far as I understood the typing syntax, it should be fine to have `Series` there, as long as anything returned is a subclass of it. If it's supposed to be changed to `Self`, that would apply to many other methods as well.

File: py-polars/polars/internals/dataframe/frame.py [hunk 3]
```diff
@@ -1275,7 +1287,7 @@ def _div(self, other: Any, floordiv: bool) -> Self:
 
         elif not isinstance(other, DataFrame):
             s = _prepare_other_arg(other, length=len(self))
-            other = DataFrame([s.rename(f"n{i}") for i in range(len(self.columns))])
+            other = type(self)([s.rename(f"n{i}") for i in range(len(self.columns))])
```
[the remark this answers is not in the document]
Comment (ml31415) [reply py-polars/polars/internals/dataframe/frame.py:1290]: It's the standard way of accessing a type from inside of the class definition, less error prone than `self.__class__` and IMHO rather commonly used. I'd rather argue, that using the class name to access the type without good reason (like in the instance check above) is a violation of the clean code principle 'don't repeat yourself'. So I don't see anything to improve here.

File: py-polars/polars/internals/series/series.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kylebarron) [reply py-polars/polars/internals/series/series.py:445]: In general with these type of methods, if you pass in a specific subclass of a `Series`, you probably want to get that exact same subclass type back. `Self` type exists to support this: https://peps.python.org/pep-0673/

In this specific case, not sure how that interacts with `other` and what type `other` is allowed to be.

**Comment by stinodego** [comment]
@ml31415 I've briefly discussed with Ritchie, and I don't really want to go ahead with this until we have a clearer vision on how we want to approach interoperability. So let's put a pin in this for now, and we'll get back to issue #7200 in the near future.

**Comment by ritchie46** [comment]
I will close this as I don't think we should have inheritance on our types. Discussion can continue in https://github.com/pola-rs/polars/issues/7200

**Pull request closed without merging**